server/api/db/sqldb/db.py

```diff
@@ -227,21 +227,48 @@ def store_run(
                 start_time=run_start_time(run_data) or now,
                 requested_logs=False,
             )
-        self._ensure_run_name_on_update(run, run_data)
-        labels = run_labels(run_data)
-        self._update_run_state(run, run_data)
-        update_labels(run, labels)
-        # Note that this code basically allowing anyone to override the run's start time after it was already set
-        # This is done to enable the context initialization to set the start time to when the user's code actually
-        # started running, and not when the run record was initially created (happening when triggering the job)
-        # In the future we might want to limit who can actually do that
-        start_time = run_start_time(run_data) or SQLDB._add_utc_timezone(run.start_time)
-        run_data.setdefault("status", {})["start_time"] = start_time.isoformat()
-        run.start_time = start_time
-        self._update_run_updated_time(run, run_data, now=now)
-        run.struct = run_data
+        self._enrich_run_model(now, run, run_data)
         self._upsert(session, [run], ignore=True)
 
+    def create_or_get_run(
+        self,
+        session: Session,
+        run_data: dict,
+        uid: str,
+        project: str = "",
+        iter: int = 0,
+    ):
+        """
+        This method is used to ensure a specific run is in the DB.
+        Due to isolation levels, it is possible that a certain session is unable to read a run from the DB since it
+        has an outdated snapshot. Here, we try to create a run, if we get a conflict, the session was rollbacked, and
+        we can now read the run from the DB.
+        """
+        logger.debug(
+            "Creating or getting run in DB",
+            project=project,
+            uid=uid,
+            iter=iter,
+            run_name=run_data["metadata"]["name"],
+        )
+        now = datetime.now(timezone.utc)
+        run = Run(
+            name=run_data["metadata"]["name"],
+            uid=uid,
+            project=project,
+            iteration=iter,
+            state=run_state(run_data),
+            start_time=run_start_time(run_data) or now,
+            requested_logs=False,
+        )
+        self._enrich_run_model(now, run, run_data)
+        try:
+            self._upsert(session, [run], silent=True)
+        except mlrun.errors.MLRunConflictError:
+            # Session was rollbacked and we now get a new snapshot
```

Comment:
I'm not sure sqlalchemy really does a rollback on an exception, I think you might need to add         `session.rollback()`

Reply:
we do the rollback in `_try_commit_obj`